screenpro/ngs/counter.py

```diff
@@ -0,0 +1,217 @@
+        # covert to polar DataFrame
+        library = pl.from_pandas(library)
+
+        self.library = library
+        
+    def _get_sgRNA_table(self):
+
+        if self.cas_type == 'cas9':
+            if self.library_type == "single_guide_design":
+                sgRNA_table = self.library.to_pandas()[['target','sgID','protospacer']].set_index('sgID')
+
+            elif self.library_type == "dual_guide_design":
+                sgRNA_table = pd.concat([
+                    self.library.to_pandas()[['target','sgID_A','protospacer_A']].rename(columns={'sgID_A':'sgID','protospacer_A':'protospacer'}),
+                    self.library.to_pandas()[['target','sgID_B', 'protospacer_B']].rename(columns={'sgID_B':'sgID','protospacer_B':'protospacer'})
+                ]).set_index('sgID')
+
+        return sgRNA_table
+
+    def get_counts_matrix(self, fastq_dir, samples,get_recombinant=False, cas_type='cas9', parallel=False, verbose=False):
+        '''Get count matrix for given samples
+        '''
+        if self.cas_type == 'cas9':
+            counts = {}
+
+            if self.library_type == "single_guide_design":
+                # TODO: Implement codes to build count matrix for given samples
+                pass
+            elif self.library_type == "dual_guide_design":
+                if get_recombinant: recombinants = {}
+                
+                # TODO: Fix the function for parallel processing
+                def process_sample(sample_id):
+                    if verbose: print(green(sample_id, ['bold']))
+                    df_count = cas9.fastq_to_count_dual_guide(
+                        R1_fastq_file_path=f'{fastq_dir}/{sample_id}_R1.fastq.gz',
+                        R2_fastq_file_path=f'{fastq_dir}/{sample_id}_R2.fastq.gz',
+                        trim5p_pos1_start=1,
+                        trim5p_pos1_length=19,
+                        trim5p_pos2_start=1,
+                        trim5p_pos2_length=19,
+                        verbose=verbose
+                    )
+                    if not get_recombinant:
+                        counts[sample_id] = cas9.map_to_library_dual_guide(
+                            df_count=df_count,
+                            library=self.library,
+                            get_recombinant=False,
+                            return_type='mapped',
+                            verbose=verbose
+                        )
+                    elif get_recombinant:
+                        cnt = cas9.map_to_library_dual_guide(
+                            df_count=df_count,
+                            library=self.library,
+                            get_recombinant=True,
+                            return_type='all',
+                            verbose=verbose
+                        )
+                        counts[sample_id] = cnt['mapped']
+                        recombinants[sample_id] = cnt['recombinant']
+                
+                if parallel:
```

Comment:
@nick-youngblut I couldn't resolve this part. Basically, I hope to submit multithread on samples to enhance the run time for `fastq2count` task.